### PR TITLE
[JUJU-3488] Add option to generate suites with minimal version condition

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Sometimes our integration tests test a recently added feature, meaning
we only wish to run the test for versions beyond some specific minor
release

Also extract wait-for-cloud-init and -prepare-integration-test out of
run-integration-test since the envvars collected by
prepare-integration-test are required for this conditional step

Add an option to juju.config which, if set, wraps our integration test
builder in a conditional clause

Implement this for the new ssh allow firewall test (test-firewall)

Also, it seems someone added some tests to the controller suite but
forgot to add them to Jenkins. Add them here as a flyby

## QA Steps

https://jenkins.juju.canonical.com/job/test-firewall-test-firewall-ssh-aws/

Verify this pipeline succeed for Juju versions on both sides of 3.2

- For 3.2 or beyond, the integration test should run
- For versions prior to 3.2, the pipeline should succeed but not run the integration test

### CODE REVIEW

File: tools/gen-wire-tests/main.go [hunk 1]
```diff
@@ -491,6 +509,23 @@ const Template = `
...
+{{- if index $.MinVersions $task_name }}
+      - conditional-step:
+          condition-kind: regex-match
+          regex: "{{ index $.MinVersions $task_name }}"
+          label: "$JUJU_VERSION"
+          steps:
+            - {{$builder}}:
+                  test_name: '{{$.SuiteName}}'
+                  setup_steps: ''
+{{- if gt (len $node.SkipTasks) 1 }}
```
Comment (nvinuesa) [inline tools/gen-wire-tests/main.go:521]: can this macro IFs be indented so it's more explicit that they are nested?

Comment (jack-w-shaw) [reply tools/gen-wire-tests/main.go]: Addressed in new commit

File: tools/gen-wire-tests/main.go [hunk 2]
```diff
@@ -55,6 +56,14 @@ var (
 	microk8s = Cloud{CloudName: "microk8s", ProviderName: "k8s"}
 )
 
+var minVersionRegex = map[string]string{
+	"3.0": "^[4-9].*|^3\\\\.([0-9]|\\\\d{2,})(\\\\.|-).*",
+	"3.1": "^[4-9].*|^3\\\\.([1-9]|\\\\d{2,})(\\\\.|-).*",
+	"3.2": "^[4-9].*|^3\\\\.([2-9]|\\\\d{2,})(\\\\.|-).*",
+	"3.3": "^[4-9].*|^3\\\\.([3-9]|\\\\d{2,})(\\\\.|-).*",
+	"3.4": "^[4-9].*|^3\\\\.([4-9]|\\\\d{2,})(\\\\.|-).*",
```
Comment (SimonRichardson) [inline tools/gen-wire-tests/main.go:64]: You definitely need to add a comment about, what this is doing. Also, I'm not sure you need `[4-9].*|` there. We're some way away from releasing a 4. version and I hope we're not using this by then. Removing that will make the regex cleaner and easier to read, esp. with all the back slashes.

**Review by SimonRichardson — APPROVED** [review]

File: tools/gen-wire-tests/main.go (hunk 2 above)
Comment (SimonRichardson) [reply tools/gen-wire-tests/main.go:72]: > I suspect that is wishful thinking

We have no option, mutlijob is not supported, and we _need_ or __require__ to get off it.

**Pull request merged**